mitsurunner.yaml

```diff
@@ -81,96 +90,172 @@ sensor:
-            // if temperature has been long enough over treshold or force defrost temperature is exceeded
-            if ( id(G_excess_min_time_passed) == true || temperature_delta >= TEMPERATURE_DELTA_TO_FORCE_DEFROST) {
-                // activate defrosting
-                id(trigger_defrosting).execute();
-                reset_min_time_of_temperature_delta = true;
-                }
-            // if excess timer were not yet passed, do nothing
-            else if ( id(start_excess_min_time_timer).is_running() ) {
-                }
-            // excess min time were not on, but other defrosting requirements were fullfilled to start it
-            else  
-                { 
-                id(start_excess_min_time_timer).execute();
-                }
-            }            
-        else {        
-            // if during the excess timer, the temperature returns back below treshold, then cancel the timer
-            reset_min_time_of_temperature_delta = true;       
-            }
-        
-        if ( reset_min_time_of_temperature_delta ) {
-            id(G_excess_min_time_passed) = false;
-            
-            // timer is needed to be stopped only if it is running
-            if ( id(start_excess_min_time_timer).is_running() ) {
-                id(start_excess_min_time_timer).stop();  
-                }
-            id(mqtt_client).publish(TOPIC_IS_EXCESS_MIN_TIME_PASSED, to_string(0.0));
-            }
 
         // if you want to get MQTT logs from temperature delta all the time, define TOPIC_TEMP_DELTA in constants.h file
         #ifdef TOPIC_TEMP_DELTA_MESSAGES
         id(mqtt_client).publish(TOPIC_TEMP_DELTA, to_string(temperature_delta));
         #endif
 
+        id(mqtt_client).publish(TOPIC_STATE, to_string(id(G_state)));
+
+        switch(id(G_state)) {
+
+            // This state is executed only in reset
+            case ST_RESET:
+                // If defrosting is going on
+                if (temperature_delta < 0.0) {
+                    id(EnterDefrostingStarted).execute();
+                }
+                // If defrosting is not going on
+                else {
+                    id(schedule_forced_defrosting).execute();
+                    id(EnterIdle).execute();
+                }
+                break;
+
+            case ST_IDLE:
+                if (temperature_delta >= TEMPERATURE_DELTA_TO_DEFROST) {
+                    id(EnterTempExceeded).execute();
+                }
+                break;
+
+            case ST_TEMP_EXCEEDED:
+                if (id(G_state_time_passed) || (temperature_delta >= TEMPERATURE_DELTA_TO_FORCE_DEFROST)) {
+                    id(EnterStartDefrosting).execute();
+                }
+                else if (temperature_delta < TEMPERATURE_DELTA_TO_DEFROST) {
+                    id(EnterIdle).execute();
+                }
+                break;
+
+            case ST_START_DEFROSTING:
+                if (temperature_delta < 0.0) {
```

Comment:
In which case the temperature delta has dropped, when start_defrozing happens?

Reply:
The purpose of this is to detect when defrosting has been really started: the heat exchanger is warmer than the outside temperature. If this does not happen before 10min timeout, we deduce that the pump has not done defrosting, and we go back to the idle state without resetting minimum heating time counter.

Reply:
I just witnessed a "fake defrosting" when the logic thought that it's time to defrost, but the pump did not defrost because the heat exchanger was still over -1°C. MitsuRunner noticed that defrosting did not start, so it went back to the idle state, as it should.

Reply:
Great, so this is new functionality to make system more intelligent regarding heatpump status.